drivers/flash/soc_flash_b91.c

```diff
@@ -109,9 +140,30 @@ static int flash_b91_write(const struct device *dev, off_t offset,
 		return -EACCES;
 	}
 
+	/* need to store data in intermediate RAM buffer in case from flash to flash write */
+	if (((uint32_t)data >= FLASH_ORIGIN) &&
+		((uint32_t)data < (FLASH_ORIGIN + FLASH_SIZE))) {
+
+		buf = k_malloc(len);
+		if (buf == NULL) {
+			return -ENOMEM;
```

Comment:
Hmm, actually after this move we can exit here with the mutex locked, so it's still a problem

Reply:
Hmm, you are right.
Fixed.